open_spiel/python/algorithms/alpha_zero.py

```diff
@@ -0,0 +1,478 @@
+    tf.variables_initializer(optim.variables())
+    tf.train.get_or_create_global_step().assign(0)
+
+    num_epoch_iters = math.ceil(len(self.replay_buffer) / float(batch_size))
+    losses = []
+    for epoch in range(num_training_epochs):
+      epoch_losses = []
+      for _ in range(num_epoch_iters):
+        train_data = self.replay_buffer.sample(batch_size, replace=True)
+        epoch_losses.append(self.bot.evaluator.update(train_data))
+
+      losses.append(epoch_losses)
+      if verbose:
+        self._print_mean_epoch_losses(epoch, epoch_losses)
+
+    return losses
+
+  def self_play(self, num_self_play_games=5000):
+    """Uses the current state of the net with MCTS to play full games against  
+
+    Args:
+      num_self_play_games: the number of self-play games to play using the 
+        current net and MCTS.
+    """
+    for _ in range(num_self_play_games):
+      self._self_play_single()
+
+  def _self_play_single(self):
+    state = self.game.new_initial_state()
+    policy_targets, state_features = [], []
+
+    while not state.is_terminal():
+      root_node = self.bot.mcts_search(state)
+      state_features.append(self.bot.evaluator.feature_extractor(state))
+      target_policy = np.zeros(self.game.num_distinct_actions(),
+                               dtype=np.float32)
+      for child in root_node.children:
+        target_policy[child.action] = child.explore_count
+      target_policy /= sum(target_policy)
+      policy_targets.append(target_policy)
+
+      action = self._select_action(root_node.children, len(state.history()))
+      state.apply_action(action)
+
+    terminal_rewards = state.rewards()
+    for i, (feature, pol) in enumerate(zip(state_features, policy_targets)):
+      value = terminal_rewards[i % 2]
+      self.replay_buffer.add(
+          MCTSResult(state_feature=feature,
+                     target_policy=pol,
+                     target_value=value))
+
+  def _select_action(self, children, game_history_len):
+    explore_counts = [(child.explore_count, child.action) for child in children]
+    if game_history_len < self.action_selection_transition:
+      probs = np_softmax(np.array([i[0] for i in explore_counts]))
+      action_index = np.random.choice(range(len(probs)), p=probs)
+      action = explore_counts[action_index][1]
+    else:
+      _, action = max(explore_counts)
+    return action
+
+  def _print_mean_epoch_losses(self, epoch, losses):
+    total_loss, policy_loss, value_loss, l2_loss = 0, 0, 0, 0
+    for l in losses:
+      t, p, v, l2 = l
+      total_loss += t
+      policy_loss += p
+      value_loss += v
+      l2_loss += l2
+
+    loss_str = "Epoch {0} mean losses. Total: {1:.3g}, Policy: {2:.3g}, Value: {3:.3g}, L2: {4:.3g}"
+    n = float(len(losses))
+    print(
+        loss_str.format(epoch, total_loss / n, policy_loss / n, value_loss / n,
+                        l2_loss / n))
+
+
+def alpha_zero_ucb_score(child, parent_explore_count, params):
```

Comment:
This was used for Chess and Go in the AlphaZero paper, but I found it worked far less well for tic-tac-toe. I originally had this as the default UCB score method, but changed it due to the tic-tac-toe experiments. Should probably switch back to using this as the default, as it was used in the paper/pseudocode.

Reply:
MCTSBot already supports uct and puct.

Reply:
It does, but not the AlphaZero variant (unless I'm missing something). For example, the AlphaZero variant has two hyperparameters, rather than just the one in `mcts.py`. From the pseudocode,
```
def ucb_score(config: AlphaZeroConfig, parent: Node, child: Node):
  pb_c = math.log((parent.visit_count + config.pb_c_base + 1) /
                  config.pb_c_base) + config.pb_c_init
  pb_c *= math.sqrt(parent.visit_count) / (child.visit_count + 1)

  prior_score = pb_c * child.prior
  value_score = child.value()
  return prior_score + value_score
```
where `pb_c_base = 19652` and `pb_c_init = 1.25` are the defaults.